src/basic/pidref.c

```diff
@@ -114,6 +114,70 @@ void pidref_done(PidRef *pidref) {
         };
 }
 
+PidRef *pidref_free(PidRef *pidref) {
+        /* Regularly, this is an embedded structure. But sometimes we want it on the heap too */
+        if (!pidref)
+                return NULL;
+
+        pidref_done(pidref);
+        return mfree(pidref);
+}
+
+int pidref_dup(PidRef *pidref, PidRef **ret) {
+        _cleanup_close_ int dup_fd = -EBADF;
+        pid_t dup_pid = 0;
+
+        assert(ret);
+
+        /* Allocates a new PidRef on the heap, making it a copy of the specified pidref. This does not try to
+         * acquire a pifd if we don't have one yet! */
+
+        if (pidref) {
```

Comment:
I made it like this on purpose, since it makes it easy to copy a bunch of fields in a struct that way, and not treat NULL/non-NULL cases specifically. I'll add a comment about this to the sources.

Compare with our usual destructors that all are happy to take NULL, and also things like free_and_strdup() which are all cool with an invalidated input.